codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Paginators.java

```diff
@@ -0,0 +1,250 @@
+        writer.writeDocs(String.format("NextPage retrieves the next %s page.", operationSymbol.getName()));
+        writer.openBlock("func (p $P) NextPage(ctx $T, optFns ...func(*Options)) ($P, error) {", "}",
+                paginatorSymbol, contextSymbol, outputSymbol, () -> {
+                    writer.addUseImports(SmithyGoDependency.FMT);
+                    writer.openBlock("if !p.HasMorePages() {", "}", () -> {
+                        writer.write("return nil, fmt.Errorf(\"no more pages available\")");
+                    });
+                    writer.write("");
+                    writer.write("params := *p.params");
+                    writer.write("params.$L = p.nextToken", inputMember);
+                    if (pageSizeMember != null) {
+                        writer.write("params.$L = p.options.Limit", symbolProvider.toMemberName(pageSizeMember));
+                    }
+
+                    writer.write("result, err := p.client.$L(ctx, &params, optFns...)",
+                            operationSymbol.getName());
+                    writer.openBlock("if err != nil {", "}", () -> {
+                        writer.write("return nil, err");
+                    });
+                    writer.write("p.firstPage = false");
+
+                    StringBuilder nilGuard = new StringBuilder();
+                    StringBuilder outputPath = new StringBuilder("result");
+
+                    List<MemberShape> outputMemberPath = paginationInfo.getOutputTokenPath();
+                    for (int i = 0; i < outputMemberPath.size(); i++) {
+                        MemberShape memberShape = outputMemberPath.get(i);
+                        outputPath.append(".");
+                        outputPath.append(symbolProvider.toMemberName(memberShape));
+
+                        // Don't need to check the last member here as we won't dereference it
+                        if (i != outputMemberPath.size() - 1) {
+                            if (i != 0) {
+                                nilGuard.append(" && ");
+                            }
+                            nilGuard.append(outputPath);
+                            nilGuard.append(" != nil");
+                        }
+                    }
+
+                    Runnable setToken = () -> {
+                        writer.write("p.nextToken = $L", outputPath);
+                    };
+                    if (nilGuard.length() > 0) {
+                        writer.openBlock("if $L {", "}", nilGuard.toString(), setToken::run);
+                    } else {
+                        setToken.run();
+                    }
+
+                    writer.openBlock("if p.options.StopOnDuplicateToken && "
+                            + "params.$L != nil && p.nextToken != nil && "
+                            + "*params.$L == *p.nextToken {", "}", inputMember, inputMember, () -> {
+                        writer.write("p.nextToken = nil");
+                    });
```

Comment:
I'm not sure this will do the correct behavior. It looks like the `p.nextToken` will only be updated if `outputPath` value is not nil.  This would mean when `outputPath` has a nil value, `p.nextToken` won't be updated. Potentially looping for ever.

To address this i think this check could be simplified, to always set `p.nextToken` with the value of `outputPath`, then do the checks for duplicate tokens afterwards.

```go
prevToken := p.nextToken
p.nextToken = result.Foo

if p.options.StopOnDuplicateToken && prevToken != nil && p.nextToken != nil && *prevToken == *p.nextToken {
     p.nextToken = nil
}
```

Also side note. The `p.options.StopOnDuplicateToken` probably should only be generated when `nextToken` type is a `string` shape. We don't want to get into comparing list/maps like in DynamoDB's paginators.

Reply:
Well `outputPath` can be nested in the response shape so you must check for nils to avoid panics. The solution here would be if you can't traverse to the token location in the path then just nil out next token to mark the paging as done.

Reply:
Well there is a couple of things here, yes you are correct about the corner case here for StopDuplicateToken. But there is a corner case for there being a member that is nil along the path, in that case we need to set next token to nil as well.

Reply:
As an example here is a case where the token is nested in a member of the output structure:
```go
prevToken := p.nextToken
	if result.EngineDefaults != nil {
		p.nextToken = result.EngineDefaults.Marker
	} else {
		p.nextToken = nil
	}

	if p.options.StopOnDuplicateToken && prevToken != nil && p.nextToken != nil && *prevToken == *p.nextToken {
		p.nextToken = nil
	}
```

Here is the simple case where we don't need to have a guard check:
```go
	prevToken := p.nextToken
	p.nextToken = result.Marker
	if p.options.StopOnDuplicateToken && prevToken != nil && p.nextToken != nil && *prevToken == *p.nextToken {
		p.nextToken = nil
	}
```

Reply:
Ah, good point, didn't see the case where there were multiple levels of members! I guess the alternative would be to always set `p.nextToken` to nil up front, but the else is fine.

Reply:
Ah yeah I can make the change to nil it up front after storing the reference, removes the else branch and cleans it up a bit.